source/opt/ssa_rewrite_pass.h

```diff
@@ -0,0 +1,237 @@
+  // block |bb|.  This continues looking up the CFG predecessors until it
+  // eventually needs to insert a Phi instruction.
+  //
+  // It returns the value for |var_id| from the RHS of the current reaching
+  // definition for |var_id|.
+  uint32_t ReadVariableRecursive(uint32_t var_id, ir::BasicBlock* bb);
+
+  // Creates a Phi candidate instruction for variable |var_id| in basic block
+  // |bb|.  Phi candidates are potential Phi instructions that will be created
+  // at the end of the SSA rewrite pass.
+  //
+  // Since the rewriting algorithm may remove Phi candidates when it finds them
+  // to be trivial, we avoid the expense of creating actual Phi instructions by
+  // keeping a pool of Phi candidates (|phi_candidates_|) during rewriting.
+  //
+  // Once the candidate Phi is created, it returns its ID.
+  PhiCandidate& CreatePhiCandidate(uint32_t var_id, ir::BasicBlock* bb);
+
+  // Adds operands to the Phi candidate |phi_cand|.  This may result in
+  // |phi_cand| becoming a trivial Phi instruction.  When that happens,
+  // |phi_cand| will get replaced by one of its operands.  This function will
+  // return either the ID for |phi_cand| or the ID of the operand that replaced
+  // it.
+  uint32_t AddPhiOperands(PhiCandidate* phi_cand);
+
+  // Attempts to remove a trivial Phi candidate |phi_cand|. Trivial Phis are
+  // those that only reference themselves and one other value |val| any number
+  // of times. This will try to remove any other Phis that become trivial after
+  // |phi_cand| is removed.
+  //
+  // If |phi_cand| is trivial, it returns the SSA ID for the value that should
+  // replace it.  Otherwise, it returns the SSA ID for |phi_cand|.
+  uint32_t TryRemoveTrivialPhi(PhiCandidate* phi_cand);
+
+  // Completes |phi_cand| by adding missing arguments or replacing unknown
+  // arguments with their correct reaching definition.
+  void CompletePhiCandidate(PhiCandidate* phi_cand);
+
+  // Finalizes processing of Phi candidates.  Once the whole function has been
+  // scanned for loads and stores, the CFG will still have some incomplete and
+  // trivial Phis.  This will add missing arguments and remove trivial Phi
+  // candidates.
+  void FinalizePhiCandidates();
+
+  // Map holding the value of every SSA-target variable at every basic block
+  // where the variable is stored. defs_at_block_[block][var_id] = val_id
+  // means that there is a store instruction for variable |var_id| at basic
+  // block |block| with value |val_id|.
+  BlockDefsMap defs_at_block_;
+
+  // Map, indexed by basic block, holding all the Phi candidates created at that
+  // basic block during SSA rewriting.  phi_candidates_[block] returns a map,
+  // indexed by variable ID, for the Phi candidate created for that variable at
+  // |block|.
+  BlockPhisMap phi_candidates_;
+
+  // SSA replacement table.  This maps variable IDs, resulting from a load
+  // operation, to the value IDs that will replace them after SSA rewriting.
```

Comment:
What do you mean by "variable IDs"?  I think it is suppose to be the result of an OpVariable instruction.  I think a better way of saying what I think this means is "This maps the result ids of load instruction to the value ids that will replace them."